src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -410,12 +382,45 @@ RouterExecutorRun(QueryDesc *queryDesc, ScanDirection direction, long count)
 								   (int) operation)));
 		}
 
-		resultsOK = ExecuteTaskAndStoreResults(queryDesc, task,
-											   isModificationQuery,
-											   sendTuples);
-		if (!resultsOK)
+		if (requiresMasterEvaluation)
 		{
-			ereport(ERROR, (errmsg("could not receive query results")));
+			ListCell *taskCell = NULL;
+			Query *query = workerJob->jobQuery;
+			Oid relationId = ((RangeTblEntry *) linitial(query->rtable))->relid;
+
+			ExecuteMasterEvaluableFunctions(query);
```

Comment:
Hmm, I think this doesn't work with the approach we implemented for INSERT/SELECT queries.

In INSERT/SELECT case, each task consists of deparsed queries such that the SELECT part of the query has been updated via `UpdateRelationNames()` (ie. each query is different and we cannot call it here again since we don't have the restriction context in the executor)